src/inputfetcher.h

```diff
@@ -0,0 +1,198 @@
+    std::counting_semaphore<> m_complete_semaphore{0};
+    std::atomic<bool> m_request_stop{false};
+
+    void ThreadLoop(size_t thread_index) noexcept
+    {
+        util::ThreadRename(strprintf("inputfetch.%i", thread_index));
+
+        while (true) {
+            m_start_semaphore.acquire();
+            if (m_request_stop.load(std::memory_order_relaxed)) {
+                return;
+            }
+            Work(thread_index);
+            m_complete_semaphore.release();
+        }
+    }
+
+    void Work(size_t thread_index) noexcept
+    {
+        try {
+            while (true) {
+                const auto input_index{m_input_counter.fetch_add(1, std::memory_order_relaxed)};
+                if (input_index >= m_inputs.size()) {
+                    return;
+                }
+                const auto [tx_index, vin_index] = m_inputs[input_index];
+                const auto& outpoint{m_block->vtx[tx_index]->vin[vin_index].prevout};
+                // If an input spends an outpoint from earlier in the block,
+                // it won't be in the cache yet but it also won't be in the db either.
+                if (m_txids.contains(outpoint.hash)) {
+                    continue;
+                }
+                if (m_cache->HaveCoinInCache(outpoint)) {
+                    continue;
+                }
+                if (auto coin{m_db->GetCoin(outpoint)}; coin) {
+                    m_coins[thread_index].emplace_back(outpoint, std::move(*coin));
+                } else {
+                    // Missing an input. This block will fail validation.
+                    // Skip remaining inputs.
+                    m_input_counter.store(m_inputs.size(), std::memory_order_relaxed);
+                    return;
+                }
+            }
+        } catch (const std::runtime_error& e) {
+            // Database error. This will be handled later in validation.
+            // Skip remaining inputs.
+            LogPrintLevel(BCLog::VALIDATION, BCLog::Level::Warning, "InputFetcher failed to fetch input: %s.\n", e.what());
+            m_input_counter.store(m_inputs.size(), std::memory_order_relaxed);
+        }
+    }
+
+public:
+
+    explicit InputFetcher(size_t worker_thread_count) noexcept
+    {
+        if (worker_thread_count == 0) {
+            return;
+        }
+        for (size_t n{0}; n < worker_thread_count; ++n) {
+            m_coins.emplace_back();
+            m_worker_threads.emplace_back(std::bind(&InputFetcher::ThreadLoop, this, n));
+        }
+        // One more coins vector for the main thread
+        m_coins.emplace_back();
+    }
+
+    //! Fetch all block inputs from db, and insert into cache.
+    void FetchInputs(CCoinsViewCache& cache, const CCoinsView& db, const CBlock& block) noexcept
+    {
+        if (block.vtx.size() <= 1 || m_worker_threads.size() == 0) {
+            return;
+        }
+
+        m_db = &db;
+        m_cache = &cache;
+        m_block = &block;
```

Comment:
As mentioned before, I really dislike these lines, a fetcher shouldn't change the internal state (especially since they're const). Can we pass these in parameters instead?